extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/caffeine/CaffeineCacheImpl.java

```diff
@@ -94,6 +96,12 @@ public CompletionStage<V> get() {
                 });
     }
 
+    @Override
+    public <K, V> Uni<V> getAsync(K key, Function<K, Uni<V>> valueLoader) {
+        Objects.requireNonNull(key, NULL_KEYS_NOT_SUPPORTED_MSG);
+        return recordStats ? computeWithStats(key, valueLoader) : computeWithoutStats(key, valueLoader);
```

Comment:
When stats are enabled we need to use `ConcurrentMap.compute()` (in order to call `statsCounter.recordHits(1)`) which more costly compared to `ConcurrentMap.computeIfAbsent()` because the remapping function is always called and the returned value is replaced.

Reply:
Okay, so that's for performance/cost purposes. That's what I guessed but I wanted to make sure 😃

Reply:
Yes, I will add a comment to make it clear.

Reply:
Caffeine uses single element arrays to capture the result of the map computations for post-processing. Then you can detect if the computation was performed and react accordingly. This would avoid the performance problem and code duplication.

Reply:
Thanks for the tip @ben-manes. Is the value loader guaranteed to be invoked on the same thread? I mean if it's not then does the JMM make guarantee that a value written inside the function will be visible?

Reply:
yep, that is exactly right. It's performed by the caller if absent through a `ConcurrentHashMap` computation. In this case the Map computation is returning a future as the mapping via the valueLoader, which does its own complex logic. The `absent` flag is still scoped only to the calling thread so it is safe, but it would not be safe if the valueLoader's future was setting it instead as that could be on a different thread.

Reply:
Done: https://github.com/quarkusio/quarkus/pull/31755